Src/VimCore/CommandUtil.fs

```diff
@@ -1087,20 +1090,134 @@ type internal CommandUtil
         BufferGraphUtil.MapSpanDownToSingle _bufferGraph span x.CurrentSnapshot
 
     /// Extend the selection for a mouse click
-    member x.ExtendSelectionForMouseDrag () =
+    member x.ExtendSelectionForMouseDrag (visualSpan: VisualSpan) =
+
+        /// Change the anchor point by switching to the appropriate visual mode
+        /// with a modified visual selection
+        let changeAnchorPoint (anchorPoint: SnapshotPoint) =
+            let visualKind = VisualKind.Character
+            let caretPoint = x.CaretPoint
+            let tabStop = _localSettings.TabStop
+            let visualSelection =
+                VisualSelection.CreateForPoints visualKind anchorPoint caretPoint tabStop
+            let argument = ModeArgument.InitialVisualSelection (visualSelection, None)
+            let modeKind =
+                if Util.IsFlagSet _globalSettings.SelectModeOptions SelectModeOptions.Mouse then
+                    ModeKind.SelectCharacter
+                else
+                    ModeKind.VisualCharacter
+            x.SwitchMode modeKind argument
+
+        /// Whether the specified point is at a word boundary
+        let isWordBoundary (point: SnapshotPoint) =
+
+            // Note that this function is safe to use in the presence of
+            // surrogate pairs because all surrogate pairs (and the code point
+            // they represent when combined) are non-word non-whitespace
+            // characters.
+            let isStart = SnapshotPointUtil.IsStartPoint point
+            let isEnd = SnapshotPointUtil.IsEndPoint point
+            if isStart || isEnd then
+                true
+            else
+                let pointChar =
+                    point
+                    |> SnapshotPointUtil.GetChar
+                let previousPointChar =
+                    point
+                    |> SnapshotPointUtil.SubtractOne
+                    |> SnapshotPointUtil.GetChar
+                let isWhite = CharUtil.IsWhiteSpace pointChar
+                let wasWhite = CharUtil.IsWhiteSpace previousPointChar
+                let isWord = TextUtil.IsWordChar WordKind.NormalWord pointChar
+                let wasWord = TextUtil.IsWordChar WordKind.NormalWord previousPointChar
+                let isEmptyLine =

```

Comment:
Consider adding a helper to `SnapshotPointUtil` that detects if the point is on an empty line.  I'm guessing that would be helpful elsewhere too.